### PR TITLE
Allows easy hosting of server side title screens [Ready]

### PR DESCRIPTION
**Pull request by Incoming5643**
Creates a system to allow server owners to place and maintain their own title screens separately from the codebase and allow for shuffled, map specific, and rare meme titles screens (see the readme for details).

Allows the tg servers to do things like #24544 without forcing the codebase to muddle our licensing even more. Offered as an alternative.

Essentially the idea is that the title screen becomes tied to a specific images folder in the config folder that is controlled by server owners instead of us. They can then run images of different licenses locally without requiring us to change our license. A license.txt is provided to allow server owners to self report the licensing of the images they use.

:cl:
imageadd: Server Owners: There is a new system for title screens accessible from config/title_screen folder.
add: This system allows for multiple rotating title screens as well as map specific title screens.
add: It also allows for hosting title screens in formats other than DMI.
add: See the readme.txt in config/title_screen for full details.
remove: The previous method of title screen selection, the define TITLESCREEN, has been depreciated by this change.
/:cl:

Contents of README.txt
---
The enclosed images folder holds the image files used as the title screen for the game. All common formats such as PNG, JPG, and GIF are supported.
Byond's DMI format is also supported, but if you use a DMI only include one image per file and do not give it an icon_state (the text label below the image).

Keep in mind that the area a title screen fills is a 480px square so you should scale/crop source images to these dimensions first.
The game won't scale these images for you, so smaller images will not fill the screen and larger ones will be cut off.

Using unnecessarily huge images can cause client side lag and should be avoided. Extremely large GIFs should preferentially be converted to DMIs.
Placing non-image files in the images folder can cause errors.

You may add as many title screens as you like, if there is more than one a random screen is chosen (see name conventions for specifics).

---

Naming Conventions:

Every title screen you add must have a unique name. It is allowed to name two things the same if they have different file types, but this should be discouraged.
Avoid using the plus sign "+" and the period "." in names, as these are used internally to classify images.

Common Titles:

Common titles are in the rotation to be displayed all the time. Any name that does not include the character "+" is considered a common title.

An example of a common title name is "clown".

The common title screen named "default" is special. It is only used if no other titles are available. You can overwrite "default" safely, but you 
should have a title named "default" somewhere in your DMI file if you don't have any other common titles. Because default only runs in the 
absence of other titles, if you want it to also appear in the general rotation you must rename it.

The common title screen named "blank.png" is also special. It is only used to fill space while the real title screen loads. You should leave this file alone.

Map Titles:

Map titles are tied to a specific in game map. To make a map title you format the name like this "(name of a map)+(name of your title)"

The spelling of the map name is important. It must match exactly the define MAP_NAME found in the relevant .DM file in the /_maps folder in 
the root directory. It can also be seen in game in the status menu. Note that there are no spaces between the two names.

It is absolutely fine to have more than one title tied to the same map.

An example of a map title name is "Omegastation+splash".

Rare Titles:

Rare titles are a just for fun feature where they will only have a 1% chance of appear in in the title screen pool of a given round.
Add the phrase "rare+" to the beginning of the name. Again note there are no spaces. A title cannot be rare title and a map title at the same time.

An example of a rare title name is "rare+explosion"

### CODE REVIEW

File: code/_onclick/hud/screen_objects.dm
Comment (Incoming5643) [inline code/_onclick/hud/screen_objects.dm]: Doesn't work

File: code/game/turfs/closed.dm
Comment (Incoming5643) [inline code/game/turfs/closed.dm]: Doesn't work either

**Comment by optimumtact** [comment]
TODo update the server end/jump animations @Cyberboss did to use the configured titlescreen

**Comment by coiax** [comment]
@optimumtact if it's randomly chosen each round, then there might still be a jump

**Comment by optimumtact** [comment]
What

**Comment by optimumtact** [comment]
or add end_screens for when someone leaves :3

**Comment by Cyberboss** [comment]
The fade screens are built to be map specific so they can transition from each other seemlessly

**Comment by Cyberboss** [comment]
Once #24171 is merged they are loaded using the map config

**Comment by Incoming5643** [comment]
Pull updated, everything works now!

**Comment by AnturK** [comment]
I think having these as simple pngs instead of having to edit dmi would be nice, not sure if it wouldn't need some esource sending changes though.

**Comment by lzimann** [comment]
I don't really like the way this is handled here. Sure, it allows for easy title swap, but doesn't seem give much control over map specific title screens.

**Comment by Incoming5643** [comment]
@AnturK I tried, it didn't want to work.

File: code/controllers/subsystem/title_screen.dm
Comment (Cyberboss) [inline code/controllers/subsystem/title_screen.dm]: You need to call the base

File: code/_onclick/hud/screen_objects.dm
Comment (Cyberboss) [inline code/_onclick/hud/screen_objects.dm]: I don't like that you're depending on the splashscreen turfs existing. Store the icon_state in SStitle

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
Also, you didn't just move the dmi, what changed?

File: code/_onclick/hud/screen_objects.dm
Comment (Incoming5643) [inline code/_onclick/hud/screen_objects.dm]: As an indestructible turf on z =2 its about as safe a bet as a turf gets. Using it as storage is a quick and easy way to make the splash screen match the end of round fade out.

Regardless I'd still be storing the turf either way because it saves a search for it later.

File: code/controllers/subsystem/title_screen.dm
Comment (Incoming5643) [inline code/controllers/subsystem/title_screen.dm]: The base what exactly? This does work currently.

**Comment by Incoming5643** [comment]
The dmi has an new iconstate without a name, a pure black screen to fill the space if anyone manages to join the server in the milliseconds before the subsystem initializes.

@AnturK I'll give it another shot

**Comment by Incoming5643** [comment]
Don't merge this in the next hour or two I'm on the verge of a better solution

**Comment by Incoming5643** [comment]
Alright now all you have to do is put image files (not just dmis) into the config/title_screens/images folder and you're all set!

**Comment by Incoming5643** [comment]
Fixed Conflicts

**Comment by AnturK** [comment]
Remove the titlescreen defines from the map definitions and should be good.

**Comment by Incoming5643** [comment]
Totally forgot about that one, thanks. Should be good now @AnturK

**Comment by Cyberboss** [comment]
HE DIDN'T CALL THE BASE KOR!!!!!!!

**Comment by Incoming5643** [comment]
You never told me what "calling the base" actually meant. I would have acted on it had I known.

I figured that after the first rewrite I must have "called the base" because you never clarified what "calling the base" meant after that point.

Sorry for conflicts all the same.

**Comment by RemieRichards** [comment]
please learn basic terminology

**Comment by Incoming5643** [comment]
At the end of the day this is a hobbyist project. If someone's unaware of a particular bit of lingo it's not really a huge ask to get some clarification.

Kinda crazy I've been doing this for nearly five years and it's never come up though.

**Comment by optimumtact** [comment]
I don't know what he means by that and I"m a professional with a degree. More importantly the only message from @Cyberboss from before the PR was merged and got hidden by getting changed code and it still is not clear.

Clear communication should avoid terminology wherever possible

**Comment by Incoming5643** [comment]
Note to future explorers: "calling the base" is just ..(), which is what I usually just call "running the parent's version of the code" or "calling the parent". It's not as catchy I admit.

Has no effect in the context of this pull + master because datum/subsystem doesn't have any special Initialize() as far as I can tell but must in cyberbosses dynamic loading station refactor.

Mostly putting the finger of blame on github for stashing away the comment somewhere far away from anyone who could have pointed that fact out to me.

**Pull request merged**
